### PR TITLE
Added Native-specific checker for properties of top level singletons

### PR DESCRIPTION
**Pull request by LepilkinaElena**

### CODE REVIEW

File: compiler/testData/diagnostics/nativeTests/topLevelSingleton.kt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// FILE: test.kt
+class AtomicInt(var value: Int)
+object Foo {
+    <!VARIABLE_IN_TOP_LEVEL_SINGLETON_WITHOUT_THERAD_LOCAL!>var field1: Int = 10<!>
+    val backer2 = AtomicInt(0)
+    var field2: Int
+        get() = backer2.value
+        set(value: Int) { backer2.value = value }
+}
```
Comment (SvyatoslavScherbina) [inline compiler/testData/diagnostics/nativeTests/topLevelSingleton.kt]: No tests for `@ThreadLocal` singletons.
Also no tests for plain top-level properties.

File: native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/ErrorsNative.kt [hunk 2]
```diff
@@ -24,6 +24,8 @@ object ErrorsNative {
     val INAPPLICABLE_SHARED_IMMUTABLE_PROPERTY = DiagnosticFactory0.create<KtElement>(Severity.ERROR)
     @JvmField
     val INAPPLICABLE_SHARED_IMMUTABLE_TOP_LEVEL = DiagnosticFactory0.create<KtElement>(Severity.ERROR)
+    @JvmField
+    val VARIABLE_IN_TOP_LEVEL_SINGLETON_WITHOUT_THERAD_LOCAL = DiagnosticFactory0.create<KtElement>(Severity.WARNING)
```
Comment (SvyatoslavScherbina) [inline native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/ErrorsNative.kt]: Has no message assigned.

Comment (LepilkinaElena) [reply native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/ErrorsNative.kt]: Sorry, seems to be lost during rebasing

File: native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeTopLevelSingletonChecker.kt [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import org.jetbrains.kotlin.psi.KtDeclaration
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.resolve.checkers.DeclarationChecker
+import org.jetbrains.kotlin.resolve.checkers.DeclarationCheckerContext
+import org.jetbrains.kotlin.resolve.hasBackingField
+
+object NativeTopLevelSingletonChecker : DeclarationChecker {
+    private val threadLocalFqName = FqName("kotlin.native.concurrent.ThreadLocal")
+
+    private val DeclarationDescriptor.isInsideTopLevelSingletonWithoutThreadLocal: Boolean
```
Comment (KirpichenkovPavel) [inline native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeTopLevelSingletonChecker.kt]: It seems, that companion objects may pass this check

File: native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeTopLevelSingletonChecker.kt [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        get() {
+            val parent = containingDeclaration as? ClassDescriptor
+            return parent != null && parent.kind.isSingleton &&
```
Comment (KirpichenkovPavel) [inline native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeTopLevelSingletonChecker.kt]: `isSingleton` returns true for enum entries also

File: native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeTopLevelSingletonChecker.kt [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        }
+
+    override fun check(declaration: KtDeclaration, descriptor: DeclarationDescriptor, context: DeclarationCheckerContext) {
+        if (declaration !is KtProperty || descriptor !is PropertyDescriptor ||
+            !descriptor.isInsideTopLevelSingletonWithoutThreadLocal
+        ) return
+        if (descriptor.isVar && declaration.delegate == null && descriptor.hasBackingField(context.trace.bindingContext) &&
+            descriptor.setter?.isDefault == true
+        ) {
+            context.trace.report(ErrorsNative.VARIABLE_IN_TOP_LEVEL_SINGLETON_WITHOUT_THERAD_LOCAL.on(declaration))
```
Comment (KirpichenkovPavel) [inline native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeTopLevelSingletonChecker.kt]: Typo in THREAD

File: native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeTopLevelSingletonChecker.kt (hunk 3 above)
Comment (LepilkinaElena) [reply native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeTopLevelSingletonChecker.kt]: That's ok

Comment (LepilkinaElena) [reply native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeTopLevelSingletonChecker.kt]: Added tests for companion objects

File: native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeTopLevelSingletonChecker.kt (hunk 4 above)
Comment (LepilkinaElena) [reply native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeTopLevelSingletonChecker.kt]: Thank you, added other checks for enums

File: native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/DefaultErrorMessagesNative.kt [hunk 6]
```diff
@@ -25,6 +25,9 @@ private val DIAGNOSTIC_FACTORY_TO_RENDERER by lazy {
             "@SharedImmutable is applicable only to val with backing field or to property with delegation"
         )
         put(ErrorsNative.INAPPLICABLE_SHARED_IMMUTABLE_TOP_LEVEL, "@SharedImmutable is applicable only to top level declarations")
+        put(ErrorsNative.VARIABLE_IN_TOP_LEVEL_SINGLETON_WITHOUT_THREAD_LOCAL, "Variable in top level singleton without @ThreadLocal")
```
Comment (SvyatoslavScherbina) [inline native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/DefaultErrorMessagesNative.kt]: "Top level" is redundant here (e.g. `companion object`).

File: native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/DefaultErrorMessagesNative.kt [hunk 7]
```diff
@@ -25,6 +25,9 @@ private val DIAGNOSTIC_FACTORY_TO_RENDERER by lazy {
...
+        put(ErrorsNative.ENUM_THREAD_LOCAL_INAPPLICABLE, "@ThreadLocal isn't applicable to enum classes")
+        put(ErrorsNative.VARIABLE_IN_ENUM, "Variable in enum class can't be changed after initialization")
```
Comment (SvyatoslavScherbina) [inline native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/DefaultErrorMessagesNative.kt]: Enum classes follow the same rules as `object` without `@ThreadLocal`, so I guess messages should probably be very similar in this part.

File: native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeTopLevelSingletonChecker.kt [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+            descriptor.annotations.findAnnotation(threadLocalFqName)?.let {
+                val reportLocation = DescriptorToSourceUtils.getSourceFromAnnotation(it) ?: declaration
+                context.trace.report(ErrorsNative.ENUM_THREAD_LOCAL_INAPPLICABLE.on(reportLocation))
+            }
+        }
+
+        // Check variables inside singletons.
+        if (descriptor !is PropertyDescriptor) return
+        (descriptor.containingDeclaration as? ClassDescriptor)?.let { parent ->
+            if (descriptor.isVar && DescriptorUtils.isEnumClass(parent)) {
```
Comment (SvyatoslavScherbina) [inline native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeTopLevelSingletonChecker.kt]: We should probably check here the same as below: that this property has backing field and default setter.

File: native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeTopLevelSingletonChecker.kt [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+                context.trace.report(ErrorsNative.VARIABLE_IN_ENUM.on(declaration))
+            } else if (parent.kind.isSingleton) {
```
Comment (SvyatoslavScherbina) [inline native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeTopLevelSingletonChecker.kt]: Note that enum classes aren't `.isSingleton`, but enum entries are.
It actually makes sense to handle both enums and enum entries above.

**Pull request merged**